### ISSUE TITLE
Incremental compiler StackOverflowError with name hashing

### ISSUE THREAD

**Issue by xuwei-k** [sbt/sbt#1237]
## steps

```
$ git clone git@github.com:xuwei-k/scalaz.git xuweik_scalaz
$ cd xuweik_scalaz
$ git checkout c27cd30
$ echo "sbt.version=0.13.2-RC1" > project/build.properties
$ sbt ++2.11.0-RC4 compile
```

or

> You can also reproduce it in scripted test `macro-arg-dep` (and most likely in any other test that uses macro). Simply add `scalaVersion := "2.11.0-RC3"` to the `build.scala`.
## problems

```
java.lang.StackOverflowError
    at scala.collection.TraversableLike$class.collect(TraversableLike.scala:282)
    at scala.collection.AbstractTraversable.collect(Traversable.scala:104)
    at xsbt.Compat$MacroExpansionOf$.unapply(Compat.scala:125)
    at xsbt.ExtractUsedNames.xsbt$ExtractUsedNames$$handleTreeNode$1(ExtractUsedNames.scala:86)
    at xsbt.ExtractUsedNames$$anonfun$handleMacroExpansion$1$1.apply(ExtractUsedNames.scala:58)
    at xsbt.ExtractUsedNames$$anonfun$handleMacroExpansion$1$1.apply(ExtractUsedNames.scala:58)
    at scala.reflect.internal.Trees$ForeachTreeTraverser.traverse(Trees.scala:1616)
    at scala.reflect.internal.Trees$TreeContextApiImpl.foreach(Trees.scala:106)
    at xsbt.ExtractUsedNames.handleMacroExpansion$1(ExtractUsedNames.scala:58)
    at xsbt.ExtractUsedNames.xsbt$ExtractUsedNames$$handleTreeNode$1(ExtractUsedNames.scala:88)
    at xsbt.ExtractUsedNames$$anonfun$handleMacroExpansion$1$1.apply(ExtractUsedNames.scala:58)
    at xsbt.ExtractUsedNames$$anonfun$handleMacroExpansion$1$1.apply(ExtractUsedNames.scala:58)
    at scala.reflect.internal.Trees$ForeachTreeTraverser.traverse(Trees.scala:1616)
    at scala.reflect.internal.Trees$TreeContextApiImpl.foreach(Trees.scala:106)
    at xsbt.ExtractUsedNames.handleMacroExpansion$1(ExtractUsedNames.scala:58)
    at xsbt.ExtractUsedNames.xsbt$ExtractUsedNames$$handleTreeNode$1(ExtractUsedNames.scala:88)
    at xsbt.ExtractUsedNames$$anonfun$handleMacroExpansion$1$1.apply(ExtractUsedNames.scala:58)
    at xsbt.ExtractUsedNames$$anonfun$handleMacroExpansion$1$1.apply(ExtractUsedNames.scala:58)
    at scala.reflect.internal.Trees$ForeachTreeTraverser.traverse(Trees.scala:1616)
    at scala.reflect.internal.Trees$TreeContextApiImpl.foreach(Trees.scala:106)
    at xsbt.ExtractUsedNames.handleMacroExpansion$1(ExtractUsedNames.scala:58)
    at xsbt.ExtractUsedNames.xsbt$ExtractUsedNames$$handleTreeNode$1(ExtractUsedNames.scala:88)
    at xsbt.ExtractUsedNames$$anonfun$handleMacroExpansion$1$1.apply(ExtractUsedNames.scala:58)
    at xsbt.ExtractUsedNames$$anonfun$handleMacroExpansion$1$1.apply(ExtractUsedNames.scala:58)
    at scala.reflect.internal.Trees$ForeachTreeTraverser.traverse(Trees.scala:1616)
    at scala.reflect.internal.Trees$TreeContextApiImpl.foreach(Trees.scala:106)
    at xsbt.ExtractUsedNames.handleMacroExpansion$1(ExtractUsedNames.scala:58)
    at xsbt.ExtractUsedNames.xsbt$ExtractUsedNames$$handleTreeNode$1(ExtractUsedNames.scala:88)
    at xsbt.ExtractUsedNames$$anonfun$handleMacroExpansion$1$1.apply(ExtractUsedNames.scala:58)

```
## notes
- sbt 0.13.2-RC1, sbt 0.13.5-M1,  sbt 0.13.5-M2
- enable name hashing
- Scala 2.11.0-RC1, 2.11.0-RC2, 2.11.0-RC4

https://travis-ci.org/xuwei-k/scalaz/builds/22375456

Running [blame on ExtractUsedNames.scala#L85](https://github.com/sbt/sbt/blame/0.13/compile/interface/src/main/scala/xsbt/ExtractUsedNames.scala#L85) shows `def handleTreeNode(node: Tree): Unit` that got recently merged in from #1163.

/cc @Duhemm, @gkossakowski, @xeno-by

**Comment by gkossakowski** [comment]
Thanks for the report! I can't reproduce it though. Here's what I did:
1. I checked out xuwei-k/scalaz@c27cd30ef2cae8468da25dd4a7d2ca44cdaf51db
2. I added `build.properties` with `sbt.version=0.13.2-RC1`
3. I started sbt and tried to compile everything and it succeeded.

What are the exact reproduction steps?

**Comment by eed3si9n** [comment]
@gkossakowski Kenji also mentions Scala 2.11.0-RC4.

**Comment by Duhemm** [comment]
Hi,

You can also reproduce it in scripted test `macro-arg-dep` (and most likely in any other test that uses macro). Simply add `scalaVersion := "2.11.0-RC3"` to the `build.scala`.

It looks like some macros have themselves attached as original tree, causing the infinite loop.

**Comment by xeno-by** [comment]
`It looks like some macros have themselves attached as original tree`. Could you elaborate?

**Comment by Duhemm** [comment]
For instance in [this test](https://github.com/sbt/sbt/tree/0.13/sbt/src/sbt-test/source-dependencies/macro-arg-dep), at some point the get the following tree : 

```
({
  val $u: c.universe.type = c.universe;
  val $m: $u.Mirror = c.universe.rootMirror;
  $u.TypeTag.apply[String]($m, {
    final class $typecreator1 extends TypeCreator {
      def <init>(): $typecreator1 = {
        $typecreator1.super.<init>();
        ()
      };
      def apply[U <: scala.reflect.api.Universe with Singleton]($m$untyped: scala.reflect.api.Mirror[U]): U#Type = {
        val $u: U = $m$untyped.universe;
        val $m: $u.Mirror = $m$untyped.asInstanceOf[$u.Mirror];
        $u.internal.reificationSupport.TypeRef($u.internal.reificationSupport.SingleType($u.internal.reificationSupport.ThisType($m.staticPackage("scala").asModule.moduleClass), $m.staticModule("scala.Predef")), $u.internal.reificationSupport.selectType($m.staticModule("scala.Predef").asModule.moduleClass, "String"), immutable.this.Nil)
      }
    };
    new $typecreator1()
  })
}: c.universe.WeakTypeTag[String])
```

This tree comes from the original tree :

```
`package`.this.materializeWeakTypeTag[String](c.universe)
```

Which comes from the original **same** tree :

```
`package`.this.materializeWeakTypeTag[String](c.universe)
```

And so on.

**Comment by gkossakowski** [comment]
It appears to me that this cyclic structure of original attributes of trees is a bug in Scala compiler. In particular, the 

``` scala
`package`.this.materializeWeakTypeTag[String](c.universe)
```

should not have any original tree set, right?

**Comment by eed3si9n** [comment]
+1 on rolling it back and potentially pursuing it as scalac bug while it's on RC. /cc @retronym

**Comment by xeno-by** [comment]
Hey I'll try to look into this tomorrow. This weekend and today was just a non-stop chain of stuff...

**Comment by xeno-by** [comment]
https://issues.scala-lang.org/browse/SI-8486

**Comment by dwijnand** [comment]
This is an issue for me when cross compiling to 2.10.4. I've been able to reduce this to it's minimum (imo). Please see https://gist.github.com/dwijnand/3b94e0d849e3ed2450da. And run `;clean ;+test`

**Comment by dwijnand** [comment]
Also, it's only an issue for me since upgrading from 0.13.2 to 0.13.5, so technically this is a regression? I've temporarily resolved this by disabling name hashing under 2.10.x, which is fine for now.

**Comment by Duhemm** [comment]
Thank you for the report ! I'm looking into this and will try to fix this as soon as possible.
